File: erlang/word-count/word_count.erl
-module(word_count).
-export([count/1]).

count(Text) ->
    counts(words(Text), dict:new()).

words(Text) ->
    [string:to_lower(Word) ||
        Word <- re:split(Text, "[_\\W]", [{return, list}]), Word =/= []
    ].

counts(Words, Dict) ->
    lists:foldl(fun(Word, Acc) ->
                        dict:update_counter(Word, 1, Acc)
                end,
                Dict,
                Words).

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map() 
  def count(sentence) do
    sentence
    |> String.downcase
    |> String.split(~r/[^\w-]|[_]/u, trim: true)
    |> frequencies
  end

  defp frequencies(words) do
    Enum.reduce(words, Map.new,
      fn(w, acc) ->
        Map.update(acc, w, 1, &(&1 + 1))
      end)
  end
end
